### PR TITLE
AVRO-2079: Add ability to generate Java8 native date/time classes (new)

### PR DESCRIPTION
**Pull request by pvorb**
As there was no progress on #248 for the past few months, I went ahead and resolved the merge conflicts of that PR and also fixed the remarks from @Kuroshii.

Here's a brief overview of what I've changed:

- Reorder imports in TestJava8TimeConversions.java
- Directly cast epochDays to int
- Directly cast millis to int
- Explicitly format java8 and joda time dates in test
- Rename DateTimeLogicalTypeType to DateTimeLogicalTypeImplementation
- Refactor registration of conversions
- Make DateTimeLigicalTypeImplementation field final
- Use parent POM version in Maven plugin tests
- Log warning message about unknown parameter values
- Remove unused imports

I would also suggest to rename the enum value `JAVA8` and all of its uses to `JSR310` because that's more specific to the `java.time.*` APIs introduced in Java 8 (see https://jcp.org/en/jsr/detail?id=310). I can add that change if you agree.

@Kuroshii I would appreciate if you could review this PR again, so maybe we'll have JSR-310 support in the next Avro release. Thanks!

### CODE REVIEW

File: lang/java/compiler/src/main/java/org/apache/avro/compiler/specific/SpecificCompiler.java [hunk 1]
```diff
@@ -225,6 +268,14 @@ public void setEnableDecimalLogicalType(boolean enableDecimalLogicalType) {
     this.enableDecimalLogicalType = enableDecimalLogicalType;
   }
 
+  public boolean useJodaForDateTimeLogicalTypes() {
```
Comment (lukejackson) [inline lang/java/compiler/src/main/java/org/apache/avro/compiler/specific/SpecificCompiler.java]: any reason not to expose the enum here? the two getters are not independent, and the interface essentially exposes four possible states when there are only two.

File: lang/java/maven-plugin/src/test/resources/unit/idl/pom-java8.xml [hunk 2]
```diff
@@ -0,0 +1,69 @@
...
+-->
+<project xmlns="http://maven.apache.org/POM/4.0.0" xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
+         xsi:schemaLocation="http://maven.apache.org/POM/4.0.0 http://maven.apache.org/xsd/maven-4.0.0.xsd">
+  <modelVersion>4.0.0</modelVersion>
+
+  <parent>
+    <artifactId>avro-parent</artifactId>
+    <groupId>org.apache.avro</groupId>
+    <version>1.9.0-SNAPSHOT</version>
+    <relativePath>../../../../../../../../../</relativePath>
```
Comment (lukejackson) [inline lang/java/maven-plugin/src/test/resources/unit/idl/pom-java8.xml]: relativePath should point to a specific pom.xml file, rather than a directory, here and in other pom.xml files touched by this change.

**Comment by lukejackson** [comment]
+1 on renaming to JSR310. (Note I am not an Avro committer).

File: lang/java/compiler/src/main/velocity/org/apache/avro/compiler/specific/templates/java/classic/record.vm [hunk 3]
```diff
@@ -163,9 +163,9 @@ public class ${this.mangle($schema.getName())}#if ($schema.isError()) extends or
 #if ($this.hasLogicalTypeField($schema))
   protected static final org.apache.avro.Conversions.DecimalConversion DECIMAL_CONVERSION = new org.apache.avro.Conversions.DecimalConversion();
 #if ($this.useJodaForDateTimeLogicalTypes())
```
Comment (lukejackson) [inline lang/java/compiler/src/main/velocity/org/apache/avro/compiler/specific/templates/java/classic/record.vm]: i think this needs to be refactored to use the new getDateTimeLogicalTypeImplementation() method

Comment (pvorb) [reply lang/java/compiler/src/main/velocity/org/apache/avro/compiler/specific/templates/java/classic/record.vm]: Good spot!

**Comment by DolgopolovOleg** [comment]
Sorry guys, it's really crucial feature for us. What do you think how long will it take to approve it?
Many thanks =)

File: lang/java/tools/src/main/java/org/apache/avro/tool/SpecificCompilerTool.java [hunk 4]
```diff
@@ -164,7 +181,7 @@ public String getShortDescription() {
       System.err.println("No input files found.");
     }
 
-    return fileSet.toArray((new File[fileSet.size()]));
+    return fileSet.toArray(new File[0]);
```
Comment (nandorKollar) [inline lang/java/tools/src/main/java/org/apache/avro/tool/SpecificCompilerTool.java]: Why did you change this? Since we know the size of the result array, the original code (allocating one before toArray, and passing it) is still fine isn't it?

Comment (pvorb) [reply lang/java/tools/src/main/java/org/apache/avro/tool/SpecificCompilerTool.java]: [Here is an excellent article](https://shipilev.net/blog/2016/arrays-wisdom-ancients/) that explains why nowadays it's better (despite counter intuitive) to use `toArray(new File[0])` in favor of `toArray(new File[fileSet.size()])`. The old code was not wrong, but I wanted to remove those redundant parentheses and switched to `new File[0]` because I was already editing that line. ;-)

Comment (pvorb) [reply lang/java/tools/src/main/java/org/apache/avro/tool/SpecificCompilerTool.java]: @nandorKollar If you prefer `toArray(new File[fileSet.size()])`, I can revert it of course.

**Comment by pvorb** [comment]
Recent changes:

- added a commit that replaces `JAVA8` by `JSR310` as suggested by me earlier in this thread. If anybody from the Avro project disagrees with that change, I will remove it.

- removed the runtime dependency on Guava's `Strings` class from the `AbstractAvroMojo`, which caused the archetype modules to fail.

- added a new entry in the `CHANGES.txt` describing the feature

@nandorKollar Is there anything else I can do?

**Comment by pvorb** [comment]
@nandorKollar Just saw that you want to remove `CHANGES.txt` in #310, so I removed the new entry in order to prevent a merge conflict.

**Comment by pvorb** [comment]
@nandorKollar Is there anything else I can do?

**Comment by pvorb** [comment]
@cutting Can you please have a look? Is there anything that needs to be changed before merging this request? It would be nice to get some feedback given the time me and others invested into it.

**Comment by pvorb** [comment]
This PR and [its precursor](#248) are by far the most upvoted pull requests for this project. Probably it’s worth considering to merge it?

**Comment by malcolmrobbins** [comment]
Hi pvorb,
we're really keen to get hold of Java8 time support within AVRO, even if this means creating our own fork for a while as we're about to undertake a significant project...  What would it take for me to do this ?  i.e. can I pull the changes and apply this to my own fork of the code?  Alternatively is there an AVRO release imminent to save me the trouble?

**Comment by malcolmrobbins** [comment]
i.e. can I get hold of vorb:use-java8-time for our "private use" until it becomes generally available

**Comment by rocketraman** [comment]
LOL: https://issues.apache.org/jira/browse/AVRO-2210

**Comment by pvorb** [comment]
@malcolmrobbins Feel free to do this. Don't expect an official release with this change anytime soon.

**Comment by malcolmrobbins** [comment]
Thanks for that - I took a copy of your branch and tried a mvn test install and the build failed on Apache Avro Tools with an error that indicates a dependency on avro-mapred:jar:tests could not be satisfied.  The actual error is:

[ERROR] Failed to execute goal on project avro-tools: Could not resolve dependencies for project org.apache.avro:avro-tools:jar:1.9.0-SNAPSHOT: Could not find artifact org.apache.avro:avro-mapred:jar:tests:1.9.0-SNAPSHOT -> [Help 1]

Do have any pointers as to how I can overcome this?  Thanks in advance.

**Comment by malcolmrobbins** [comment]
FYI I overcame this problem.  To do this I simply needed to run the build.sh script in the lang/java folder rather than following the more explicit mvn instructions in the How To Contribute guide.  Now I just need to figure out how to get the avro-maven-plugin to generate Java sources that use Java 8 datetime rather than Joda

**Comment by kgalieva** [comment]
@malcolmrobbins You need to add `<dateTimeLogicalTypeImplementation>jsr310</dateTimeLogicalTypeImplementation>` into `configuration` section.

@pvorb thank you! This patch works fine on our schemas!

**Comment by Fokko** [comment]
I've ran some tests against [Divolte](https://github.com/divolte/divolte-collector) and it looks great. Can we get this merged?

File: lang/java/avro/src/main/java/org/apache/avro/data/Jsr310TimeConversions.java [hunk 5]
```diff
@@ -0,0 +1,158 @@
...
+import org.apache.avro.Conversion;
+import org.apache.avro.LogicalType;
+import org.apache.avro.Schema;
+
+import java.time.Instant;
+import java.time.LocalDate;
+import java.time.LocalTime;
+import java.util.concurrent.TimeUnit;
+
+public class Jsr310TimeConversions {
```
Comment (nandorKollar) [inline lang/java/avro/src/main/java/org/apache/avro/data/Jsr310TimeConversions.java]: It would be nice to override getRecommendedSchema methods for each conversion similar as in the Joda time conversion class (TimeConversions)

File: lang/java/compiler/src/main/java/org/apache/avro/compiler/specific/SpecificCompiler.java [hunk 6]
```diff
@@ -153,16 +174,38 @@ public SpecificCompiler(Protocol protocol) {
...
+  }
+
+  public SpecificCompiler(Schema schema, DateTimeLogicalTypeImplementation dateTimeLogicalTypeImplementation) {
+    this(dateTimeLogicalTypeImplementation);
     enqueue(schema);
     this.protocol = null;
   }
 
+  /**
+   * Creates a specific compler with the default (Joda) type for date/time related logical types.
```
Comment (nandorKollar) [inline lang/java/compiler/src/main/java/org/apache/avro/compiler/specific/SpecificCompiler.java]: nit: compler -> compiler

**Comment by pvorb** [comment]
@nandorKollar I changed the code according to your requests. Good spot about overriding the `getRecommendedSchema()` methods in each conversion!

**Comment by nandorKollar** [comment]
Merged to master, thanks @aukevanleeuwen and @pvorb for working on this!

**Comment by pvorb** [comment]
Thanks, @nandorKollar!

**Comment by Blackdread** [comment]
Hi, it is merged but avro maven plugin 1.8.0 still does not support that, right?
I am trying to set my avsc file but still get jodatime once generated

**Comment by Haybu** [comment]
Any rough estimation when 1.9.0 will be released?

**Comment by Fokko** [comment]
Next week I'll cut the branch and release the RC. Please subscribe to the dev maillist so you can help test the RC 👍
